pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/ConsumptionDelayTracker.java

```diff
@@ -0,0 +1,295 @@
+    _serverMetrics.addCallbackTableGaugeIfNeeded(_tableNameWithType, ServerGauge.MAX_PINOT_CONSUMPTION_DELAY_MS,
+        () -> (long) getMaxPinotConsumptionDelay());
+  }
+
+  // Constructor that uses default timeout
+  public ConsumptionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager tableDataManager) {
+    this(serverMetrics, tableNameWithType, tableDataManager, INITIAL_TIMER_THREAD_DELAY_MS,
+        TIMER_THREAD_TICK_INTERVAL_MS);
+  }
+
+  /**
+   * Use to set or rest the aging of reported values.
+   * @param enableAging true if we want maximum to be aged as per sample time or false if we do not want to age
+   *                   samples
+   */
+  @VisibleForTesting
+  void setEnableAging(boolean enableAging) {
+    _enableAging = enableAging;
+  }
+
+  /*
+   * Called by LLRealTimeSegmentDataManagers to post delay updates to this tracker class.
+   * If the new sample represents a new Maximum we update the current maximum.
+   * If the new sample was for the partition that was maximum, but delay is not maximum anymore, we must select
+   * a new maximum.
+   *
+   * @param delayInMilliseconds pinot consumption delay being recorded.
+   * @param sampleTime sample time.
+   * @param partitionGroupId partition ID for which this delay is being recorded.
+   */
+  public void storeConsumptionDelay(long delayInMilliseconds, long sampleTime, long partitionGroupId) {
+    // Store new measure and wipe old one for this partition
+    _partitionToDelaySampleMap.put(partitionGroupId, new DelayMeasure(sampleTime, delayInMilliseconds));
+    // If we are consuming we do not need to track this partition for removal.
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+  }
+
+  /*
+   * Handle partition removal event. This must be invoked when we stop serving a given partition for
+   * this table in the current server.
+   * This function will be invoked when we receive CONSUMING -> DROPPED / OFFLINE state transitions.
+   *
+   * @param partitionGroupId partition id that we should stop tracking.
+   */
+  public void stopTrackingPartitionConsumptionDelay(long partitionGroupId) {
+    removePartitionId(partitionGroupId);
+  }
+
+  /*
+   * This method is used for timing out inactive partitions, so we don't display their metrics on current server.
+   * When the inactive time exceeds some threshold, we read from ideal state to confirm we still host the partition,
+   * if not we remove the partition from being tracked locally.
+   * This call is to be invoked by a timer thread that will periodically wake up and invoke this function.
+   */
+  public void timeoutInactivePartitions() {
+    List<Integer> partitionsHostedByThisServer = null;
+    try {
+      partitionsHostedByThisServer = getPartitionsHostedByThisServerPerIdealState();
+    } catch (Exception e) {
+      _logger.error("Failed to get partitions hosted by this server for table {}", _tableNameWithType);
```

Comment:
Dont we already set the logger to contain table name ?

Reply:
good point